src/maxwell-web-view.c

```diff
@@ -0,0 +1,897 @@
+    {
+      const gchar *uri = webkit_uri_scheme_request_get_uri (request);
+      GdkPixbuf *pixbuf = NULL;
+      gchar *pixbuf_id;
+
+      /*
+       * maxwell:///id[?pixbuf]
+       *
+       * Where 'id' is the child id and 'pixbuf' is an optional GdkPixbuf
+       * pointer encoded in base64
+       *
+       * GdkPixbuf are created in damage-event handler and pushed to priv->pixbufs
+       * for us to consume if the uri includes the pixbuf id.
+       * Otherwise if there is no specific image requested we return the whole
+       * offscreen
+       */
+      if ((pixbuf_id = g_strstr_len (uri, -1, "?")))
+        {
+          guchar *image_data;
+          gsize len;
+
+          if ((image_data = g_base64_decode (&pixbuf_id[1], &len)) &&
+              len == sizeof (GdkPixbuf *))
+            {
+              pixbuf = *((GdkPixbuf **)image_data);
+              if (!g_list_find (priv->pixbufs, pixbuf))
+                pixbuf = NULL;
+            }
+
+          g_free (image_data);
+        }
+      else
+        {
+          cairo_surface_t *surface;
+
+          /* No pixbuf id, return the whole offscreen */
+          surface = gdk_offscreen_window_get_surface (data->offscreen);
+          pixbuf = gdk_pixbuf_get_from_surface (surface, 0, 0,
+                                                data->alloc.width,
+                                                data->alloc.height);
+
+          priv->pixbufs = g_list_prepend (priv->pixbufs, pixbuf);
+        }
+
+      if (pixbuf &&
+          gdk_pixbuf_get_colorspace (pixbuf) == GDK_COLORSPACE_RGB &&
+          gdk_pixbuf_get_bits_per_sample (pixbuf) == 8 &&
+          gdk_pixbuf_get_has_alpha (pixbuf))
+        {
+          const guint8 *pixels = gdk_pixbuf_read_pixels (pixbuf);
+          GInputStream *stream = NULL;
+          gsize len;
+
+          len = gdk_pixbuf_get_height (pixbuf) *
+                gdk_pixbuf_get_rowstride (pixbuf);
+
+          stream = g_memory_input_stream_new_from_data (pixels, len, NULL);
+          webkit_uri_scheme_request_finish (request, stream, len,
+                                            "application/octet-stream");
+
+          /* Add a week reference to free the Pixbuf when stream if finalized */
+          g_object_weak_ref (G_OBJECT (stream),
+                             (GWeakNotify) g_object_unref,
+                             pixbuf);
+
+          /* Pixbuf is no longer our responsibility, stream will take care
+           * of freeing it when it's finalized
+           */
+          priv->pixbufs = g_list_remove (priv->pixbufs, pixbuf);
+
+          g_object_unref (stream);
+          return;
+        }
+      else if (pixbuf)
+        {
+          /* Pixbuf is not in RGBA format, ignore it */
+          priv->pixbufs = g_list_remove (priv->pixbufs, pixbuf);
+          g_object_unref (pixbuf);
+        }
+    }
+
+  error = g_error_new_literal (MAXWELL_ERROR, 0,
```

Comment:
Instead of `0`, I'd like an enum of maxwell error codes.  It's fine if for now you only have a generic "failed" error, but the enum value should be used here.